examples/tutorials/mnist_pytorch/idle_gpu_watcher.py

```diff
@@ -0,0 +1,176 @@
+        self._gpu_util_samples = {}  # gpu_uuid: <circular queue of samples>
+
+    def _get_nvidia_smi(self):
+        # nvidia-smi --help-query-gpu: "utilization.gpu":
+        # Percent of time over the past sample period during which one or more kernels was
+        # executing on the GPU. The sample period may be between 1 second and 1/6 second
+        # depending on the product.
+        output = {}
+
+        csv_output = subprocess.check_output(self._nvida_smi_cmd).decode()
+        for line in csv_output.splitlines():
+            match = re.match(self._line_regex, line)
+            if match is None:
+                raise RuntimeError(f"Unexpected output format: {line}")
+
+            data = match.groupdict()
+            output[data["uuid"]] = int(data["gpu_util"])
+        return output
+
+    def _check_idle(self):
+        gpu_utils = self._get_nvidia_smi()
+
+        for gpu_uuid, gpu_util in gpu_utils.items():
+            if self._gpu_util_samples.get(gpu_uuid) is None:
+                self._gpu_util_samples[gpu_uuid] = deque()
+
+            self._gpu_util_samples[gpu_uuid].append(gpu_util)
+
+            if len(self._gpu_util_samples[gpu_uuid]) < self._num_samples:
+                continue
+
+            if all(s < self._min_threshhold for s in self._gpu_util_samples[gpu_uuid]):
+                sys.stderr.write(
+                    f"IDLE_GPU_WATCHER: {gpu_uuid} usage below threshhold {self._min_threshhold}% "
+                    f"for last {self._sample_freq * self._num_samples} seconds. Killing trial.\n")
+                sys.exit(1)
+            if self._debug:
+                print(f"DEBUG: {gpu_uuid} usage {gpu_util}% ")
+
+            self._gpu_util_samples[gpu_uuid].popleft()
+
+    def watch_process(self, args):
+        if args is None or len(args) == 0:
+            raise ValueError("args cannot be emtpy")
+
+        p = subprocess.Popen(args)
+        print("IDLE_GPU_WATHCER: "
+              f"Starting idle GPU watcher: min_threshhold={self._min_threshhold}%, "
+              f"sample_freq={self._sample_freq}, num_samples={self._num_samples}, "
+              f"delay_samples={self._delay_samples}, window_size={self._full_window}")
+
+        try:
+            while True:
+                if p.poll() is None:
+                    time.sleep(self._sample_freq)
+
+                    if self._delay_samples > 0:
+                        self._delay_samples -= 1
+                        continue
+
+                    self._check_idle()
+                    continue
+
+                break
+
+        finally:
+            if p.poll() is None:
+                print("IDLE_GPU_WATCHER: Training process still alive, killing.")
+                p.kill()  # SIGKILL because it is likely hung
+
+        return p.wait()
+
+
+class AParser(argparse.ArgumentParser):
+    def error(self, message):
+        description = """
+To use with Determined >= 0.17.7, add the following to your experiment config:
```

Comment:
Maybe use [this](https://docs.python.org/3/library/argparse.html#description) instead?

Reply:
I could use this but I would also have to add FormatterClass things to get the spacing right, seems like overkill right here.